src/spawn_stubs.c

```diff
@@ -488,6 +502,188 @@ CAMLprim value spawn_unix(value v_env,
     }
   }
 
+  if (new_mask || old_mask)
+    pthread_sigmask(SIG_SETMASK, new_mask, old_mask);
+}
+
+#if defined(USE_POSIX_SPAWN)
+
+CAMLprim value spawn_unix(value v_env,
+                          value v_cwd,
+                          value v_prog,
+                          value v_argv,
+                          value v_stdin,
+                          value v_stdout,
+                          value v_stderr,
+                          value v_use_vfork,
+                          value v_setpgid,
+                          value v_sigprocmask)
+{
+  CAMLparam4(v_env, v_cwd, v_prog, v_argv);
+  CAMLlocal1(e_arg);
+  e_arg = Nothing;
+
+  pid_t pid;
+  int tmp_fds[3] = {0};  // invariant: initialized > 2
+
+  int e_error;
+  char *e_function = NULL;
+
+  posix_spawn_file_actions_t actions;
+  if (posix_spawn_file_actions_init(&actions))
+    caml_raise_out_of_memory();
+
+  posix_spawnattr_t attr;
+  if (posix_spawnattr_init(&attr))
+    caml_raise_out_of_memory();
+
+  struct spawn_info info;
+  init_spawn_info(&info,
+                  v_env, v_cwd, v_prog, v_argv,
+                  v_stdin, v_stdout, v_stderr, v_setpgid);
+  init_spawn_info_sigmask(&info, v_sigprocmask, NULL, NULL);
+
+  short attr_flags = POSIX_SPAWN_SETSIGMASK;
+  if (info.set_pgid) attr_flags |= POSIX_SPAWN_SETPGROUP;
+  e_error = posix_spawnattr_setflags(&attr, attr_flags);
+  if (e_error) {
+    e_function = "setpgid";
+    goto cleanup;
+  }
+
+  e_error = posix_spawnattr_setsigmask(&attr, &info.child_sigmask);
+  if (e_error) {
+    e_function = "pthread_sigmask";
+    goto cleanup;
+  }
+
+  if (info.set_pgid) {
+    e_error = posix_spawnattr_setpgroup(&attr, info.pgid);
+    if (e_error) {
+      e_function = "setpgid";
+      goto cleanup;
+    }
+  }
+
+  switch (info.cwd_kind) {
+    case INHERIT: break;
+    case PATH:
+      e_error = posix_spawn_file_actions_addchdir_np(&actions, info.cwd.path);
+      if (e_error) {
+        e_function = "chdir";
+        e_arg = Field(v_cwd, 0);
+        goto cleanup;
+      }
+      break;
+    case FD:
+      e_error = posix_spawn_file_actions_addfchdir_np(&actions, info.cwd.fd);
+      if (e_error) {
+        e_function = "fchdir";
+        e_arg = Field(v_cwd, 0);
+        goto cleanup;
+      }
+      break;
+  }
+
+  for (int fd = 0; fd < 3; fd++) {
+    int tmp_fd = tmp_fds[fd] = safe_dup(info.std_fds[fd]);
```

Comment:
We're dupping fds in the parent now, which I think resets close-on-exec, which means that any processes spawned concurrently (through our library or not) will inherit these fds, which seems sad.

I can't think of any way to do it race-free (dup3 doesn't seem to help because it requires us to know a free fd), but at least maybe we should set close-on-exec after dup?
(reading the doc of `posix_spawn`, it seems it should be able to deal with close-on-exec fds?)

Reply:
Crazy idea how to do it race-free (sans error handling):

```
int dup_close_on_exec(int fd) {
  int new_fd = open("/", O_CLOEXEC);
  dup3(fd, new_fd, O_CLOEXEC);
  return new_fd;
}
```

Reply:
Hmmm. Not that crazy. What do you make of this?:

```c
/* Same as [dup] but 1) ensures the result isn't a standard FD, 2) applies
   [flags] (via dup3) to the new FD, and 3) on error, if [culprit] is non-null,
   assigns to [*culprit] a string literal naming the function that failed.
*/
static int safe_dup(int fd, int flags, char const **culprit)
{
  int new_fd = flags? open("/", O_CLOEXEC) : dup(fd);

  if (new_fd < 0) {
    if (culprit) {
      *culprit = flags? "open" : "dup";
    }
  }

  else if (new_fd < 3) {
    int next = safe_dup(fd, flags, culprit);
    close(new_fd);
    new_fd = next;
  }

  else if (flags) {
    new_fd = dup3(fd, new_fd, flags);
    if (new_fd < 0 && culprit) {
      *culprit = "dup3";
    }
  }

  return new_fd;
}
```

This might be easier in OCaml, too.

Reply:
I'm not sure this function benefitted from solving every problem at the same time. I would probably still keep at least the retry loop in a separate function. Anyway, I found that `F_DUPFD_CLOEXEC` is a thing (a flag you can pass to `fcntl`). Maybe that just works, and we don't need the ugly `open`/`dup3` trick?